# Mnist data with CNN

## 1. Load libraries

In [1]:
import numpy as np
import tensorflow
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Conv2D, Flatten, MaxPooling2D
from tensorflow.keras.optimizers import RMSprop, Adam
from tensorflow.keras.datasets import mnist

## 2. Prepare data

In [2]:
(mnist_train_images, mnist_train_labels), (mnist_test_images, mnist_test_labels) = mnist.load_data()

In [3]:
mnist_train_image=mnist_train_images[0:10000,:,:]
mnist_train_label=mnist_train_labels[0:10000]
mnist_test_image=mnist_train_images[0:10000,:,:]
mnist_test_label=mnist_train_labels[0:10000]


In [4]:
from tensorflow.keras import backend as K
if K.image_data_format()=='channel_first':
    train_image=mnist_train_image.reshape(mnist_train_image.shape[0],1,28,28)
    test_image = mnist_test_image.reshape(mnist_test_image.shape[0], 1, 28, 28)
    input_shape=(1,28,28)
    
else:
    train_image=mnist_train_image.reshape(mnist_train_image.shape[0],28,28,1)
    test_image=mnist_test_image.reshape(mnist_test_image.shape[0], 28, 28, 1)
    input_shape=(28,28,1)
    
train_image=train_image.astype('float32')
test_image=test_image.astype('float32')
train_image/=255
test_image/=255
print(np.shape(train_image),np.shape(test_image))

(10000, 28, 28, 1) (10000, 28, 28, 1)


In [5]:
mnist_train_image.shape

(10000, 28, 28)

In [6]:
test_label=tensorflow.keras.utils.to_categorical(mnist_test_label,10)
train_label=tensorflow.keras.utils.to_categorical(mnist_train_label,10)
print(np.shape(test_label))
print(np.shape(train_label))

(10000, 10)
(10000, 10)


## 3. Build model

In [7]:
model=Sequential()
model.add(Conv2D(32, kernel_size=(3,3), activation='relu', input_shape=input_shape))
model.add(Conv2D(64, kernel_size=(3,3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='sigmoid'))


In [8]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 12, 12, 64)        0         
_________________________________________________________________
dropout (Dropout)            (None, 12, 12, 64)        0         
_________________________________________________________________
flatten (Flatten)            (None, 9216)              0         
_________________________________________________________________
dense (Dense)                (None, 128)               1179776   
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0

In [9]:
## Compile model
model.compile(loss='categorical_crossentropy',
             metrics=['accuracy'],
             optimizer='adam')

In [10]:
## train model
model.fit(train_image,train_label,
          batch_size=100,
         epochs=8,
         verbose=1,
         validation_data=(test_image, test_label))

Train on 10000 samples, validate on 10000 samples
Epoch 1/8
10000/10000 [==============================] - 52s 5ms/sample - loss: 0.8612 - accuracy: 0.7578 - val_loss: 0.1792 - val_accuracy: 0.9484
Epoch 2/8
10000/10000 [==============================] - 48s 5ms/sample - loss: 0.2979 - accuracy: 0.9236 - val_loss: 0.0904 - val_accuracy: 0.9740
Epoch 3/8
10000/10000 [==============================] - 45s 5ms/sample - loss: 0.1809 - accuracy: 0.9495 - val_loss: 0.0579 - val_accuracy: 0.9825
Epoch 4/8
10000/10000 [==============================] - 42s 4ms/sample - loss: 0.1421 - accuracy: 0.9584 - val_loss: 0.0502 - val_accuracy: 0.9853
Epoch 5/8
10000/10000 [==============================] - 42s 4ms/sample - loss: 0.1077 - accuracy: 0.9686 - val_loss: 0.0333 - val_accuracy: 0.9895
Epoch 6/8
10000/10000 [==============================] - 42s 4ms/sample - loss: 0.0900 - accuracy: 0.9719 - val_loss: 0.0289 - val_accuracy: 0.9902
Epoch 7/8
10000/10000 [==============================] - 40s 4

In [11]:
score = model.evaluate(test_image, test_label, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

10000/10000 [==============================] - 12s 1ms/sample - loss: 0.0149 - accuracy: 0.9958
Test loss: 0.014920969383441843
Test accuracy: 0.9958


Using CNN increased accuracy from 98% to 99.5% which is very great !!